File: examples/typescript/subscription/subscription.ts
/**
 * subscription example — TypeScript SDK.
 *
 * Freemium lifecycle: quota wall → trial grant → plan upgrade.
 *   SYNAPSE_API (default http://localhost:8000)
 *   SYNAPSE_TOKEN (access token for an org owner)
 */
import {
  SynapseClient,
  SynapseFeatureGatedError,
  SynapseLimitError,
} from "@synapse-saas/client";

const api = process.env.SYNAPSE_API ?? "http://localhost:8000";
const token = process.env.SYNAPSE_TOKEN;
if (!token) {
  console.error("Set SYNAPSE_TOKEN (login via the console first)");
  process.exit(1);
}

const orgId = process.env.SYNAPSE_ORG ?? "";
const client = new SynapseClient(api, { accessToken: token, orgId });

async function main(): Promise<void> {
  // ── Where we start: the free plan ──────────────────────────────────────
  const start = (await client.subscription.current()) as {
    subscription: { plan_snapshot: { key: string } };
    entitlements: { plan_key: string; features: string[] };
  };
  console.log(
    `plan=${start.entitlements.plan_key}, features=[${start.entitlements.features.join(", ")}]`,
  );

  // ── Hit the seat quota: 402 with machine-readable hints ───────────────
  try {
    for (let i = 0; i < 5; i++) {
      await client.members.invite(`seat-${i}@example.com`);
    }
  } catch (err) {
    if (err instanceof SynapseLimitError) {
      console.log(
        `quota wall: ${err.metric}=${err.limit} → upgrade at ${String((err.body as { upgrade_url?: string }).upgrade_url)}`,
      );
    } else {
      throw err;
    }
  }

  // ── Trial grant: a paid feature without a plan change ──────────────────
  await client.entitlements.grant("advanced_reports", "promo", { durationDays: 14 });
  const granted = (await client.entitlements.effective()) as { features: string[] };
  console.log(
    `after grant: advanced_reports=${granted.features.includes("advanced_reports")} (plan unchanged)`,
  );

  // ── A gated feature now passes; other gates still throw ───────────────
  try {
    await client.entitlements.grant("sso", "promo", { durationDays: 1 });
    console.log("sso granted too");
  } catch (err) {
    if (err instanceof SynapseFeatureGatedError) {
      console.log(`sso unavailable — available in: ${err.availableIn.join(", ")}`);
    }
  }

  // ── Plan upgrade: the cap moves ────────────────────────────────────────
  await client.subscription.change("starter");
  const after = (await client.subscription.current()) as {
    entitlements: { plan_key: string };
  };
  console.log(`after upgrade: plan=${after.entitlements.plan_key}, seats=10 — invites pass now`);
}

main().catch((err) => {
  console.error("failed:", err);
  process.exit(1);
});

File: examples/python/subscription/subscription.py
"""subscription (Python) — freemium lifecycle: quota wall → trial grant → upgrade.

SYNAPSE_API (default http://localhost:8000)
SYNAPSE_TOKEN (access token for an org owner), SYNAPSE_ORG (org uuid)
"""

from __future__ import annotations

import os
import sys

from synapse_saas_client import SynapseClient, SynapseLimitError

API = os.environ.get("SYNAPSE_API", "http://localhost:8000")
TOKEN = os.environ.get("SYNAPSE_TOKEN", "")
ORG = os.environ.get("SYNAPSE_ORG", "")

if not TOKEN:
    sys.exit("Set SYNAPSE_TOKEN and SYNAPSE_ORG (login via the console first)")


def main() -> None:
    client = SynapseClient(API, access_token=TOKEN, org_id=ORG)

    # ── Where we start: the free plan ──────────────────────────────────────
    start = client.subscription.current()
    entitlements = start["entitlements"]
    print(f"plan: {entitlements['plan_key']}, features: {entitlements['features']}")

    # ── Hit the seat quota: typed 402 with hints ──────────────────────────
    for i in range(5):
        try:
            client.members.invite(f"seat-{i}@example.com")
        except SynapseLimitError as err:
            print(f"quota wall: {err.metric}={err.limit} → upgrade at {err.body.get('upgrade_url')}")
            break

    # ── Trial grant: a paid feature without a plan change ─────────────────
    client.entitlements.grant("advanced_reports", "promo", duration_days=14)
    granted = client.entitlements.effective()
    print(
        f"after grant: advanced_reports={'advanced_reports' in granted['features']} "
        f"(plan unchanged: {granted['plan_key']})"
    )

    # ── Plan upgrade: the cap moves ────────────────────────────────────────
    client.subscription.change("starter")
    after = client.subscription.current()
    print(f"after upgrade: plan={after['entitlements']['plan_key']}, seats=10 — invites pass now")


if __name__ == "__main__":
    main()
